src/backend/distributed/metadata/node_metadata.c

```diff
@@ -753,13 +754,662 @@ PgDistTableMetadataSyncCommandList(void)
+			continue;
+		}
+		MemoryContextSwitchTo(metadataSyncContext);
+
+		if (nextTuple)
+		{
+			bool isNullArray[Natts_pg_dist_partition];
+			Datum datumArray[Natts_pg_dist_partition];
+			heap_deform_tuple(nextTuple, tupleDesc, datumArray,
+							  isNullArray);
+			Datum relationIdDatum = datumArray[Anum_pg_dist_partition_logicalrelid - 1];
+			Oid relationId = DatumGetObjectId(relationIdDatum);
+
+			if (!PartitionedTable(relationId))
+			{
+				continue;
+			}
+
+			List *partitionList = PartitionList(relationId);
+			List *detachPartitionCommands =
+				GenerateDetachPartitionCommandRelationIdList(partitionList);
+			detachPartitionCommands = lcons(DISABLE_DDL_PROPAGATION,
+											detachPartitionCommands);
+			detachPartitionCommands = lappend(detachPartitionCommands,
+											  ENABLE_DDL_PROPAGATION);
+
+			SendMetadataCommandListToWorkerListInCoordinatedTransaction(nodesWithMetadata,
+																		CurrentUserName(),
+																		detachPartitionCommands);
+		}
+	}
+
+	systable_endscan(scanDesc);
+	table_close(relation, AccessShareLock);
+	list_free(nodesWithMetadata);
+
+	/* break all sequence deps for citus tables and remove all shell tables */
+	List *commandList = NIL;
+	commandList = lappend(commandList, BREAK_CITUS_TABLE_SEQUENCE_DEPENDENCY_COMMAND);
+	commandList = lappend(commandList, REMOVE_ALL_SHELL_TABLES_COMMAND);
+
+	SendMetadataCommandListToWorkerListInCoordinatedTransaction(nodesWithMetadata,
+																CurrentUserName(),
+																commandList);
+
+	MemoryContextSwitchTo(oldContext);
+	MemoryContextDelete(metadataSyncContext);
+}
+
+
+/*
+ * SendIterableInterTableRelationshipCommands sends intertable relationship commands
+ * to workers within metadata connection.
+ */
+void
+SendIterableInterTableRelationshipCommands(List *nodeList)
+{
+	/* find relevant worker nodes */
+	List *nodesWithMetadata = NIL;
+	WorkerNode *workerNode = NULL;
+	foreach_ptr(workerNode, nodeList)
+	{
+		if (NodeIsPrimary(workerNode) && !NodeIsCoordinator(workerNode))
+		{
+			nodesWithMetadata = lappend(nodesWithMetadata, workerNode);
+		}
+	}
+
+	if (nodesWithMetadata == NIL)
+	{
+		return;
+	}
+
+	ScanKeyData scanKey[1];
+	int scanKeyCount = 0;
+
+	Relation relation = table_open(DistPartitionRelationId(), AccessShareLock);
+
+	TupleDesc tupleDesc = RelationGetDescr(relation);
+	SysScanDesc scanDesc = systable_beginscan(relation,
+											  InvalidOid, false,
+											  NULL, scanKeyCount, scanKey);
+
+	MemoryContext metadataSyncContext = AllocSetContextCreate(CurrentMemoryContext,
+															  "metadata sync context",
+															  ALLOCSET_DEFAULT_SIZES);
+	MemoryContext oldContext = MemoryContextSwitchTo(metadataSyncContext);
+
+	bool scanEnded = false;
+	HeapTuple nextTuple = NULL;
+	while (!scanEnded)
+	{
+		MemoryContextReset(metadataSyncContext);
+
+		MemoryContextSwitchTo(oldContext);
+		nextTuple = systable_getnext(scanDesc);
+		if (!HeapTupleIsValid(nextTuple))
+		{
+			scanEnded = true;
```

Comment:
do we really need both nextTuple and scanEnded variables? Can't we reduce it to a single variable?

Reply:
I also removed scanEnded in the next commit.